confluence-slack-integration/confluence-8-compat/src/main/java/com/atlassian/confluence/plugins/slack/util/compat/Confluence8CompatibilityHandler.java

```diff
@@ -0,0 +1,59 @@
+package com.atlassian.confluence.plugins.slack.util.compat;
+
+import com.atlassian.confluence.search.v2.ContentSearch;
+import com.atlassian.confluence.search.v2.ISearch;
+import com.atlassian.confluence.search.v2.SearchQuery;
+import com.atlassian.confluence.search.v2.SpacePermissionQueryFactory;
+import com.atlassian.confluence.search.v2.query.ArchivedSpacesQuery;
+import com.atlassian.confluence.search.v2.query.BooleanQuery;
+import com.atlassian.confluence.search.v2.query.ContentPermissionsQuery;
+import com.atlassian.confluence.search.v2.query.TextQuery;
+import com.atlassian.confluence.search.v2.sort.RelevanceSort;
+import com.atlassian.confluence.user.ConfluenceUser;
+import com.atlassian.plugin.spring.scanner.annotation.imports.ComponentImport;
+import com.atlassian.sal.api.component.ComponentLocator;
+
+import javax.annotation.Nullable;
+import java.util.HashSet;
+import java.util.Set;
+
+public class Confluence8CompatibilityHandler extends BaseConfluenceCompatibilityHandler
+        implements ConfluenceCompatibilityHandler {
+    public static final Confluence8CompatibilityHandler INSTANCE = new Confluence8CompatibilityHandler();
+
+    // component is imported here instead of ComponentImports class intentionally because SpacePermissionQueryFactory
+    // is only available since Confluence 7.17, while ComponentImports are compiled against libraries of older Confluence
+    @ComponentImport
+    private SpacePermissionQueryFactory spacePermissionQueryFactory;
+
+    private Confluence8CompatibilityHandler() {}
+
+    @Override
+    public ISearch buildSearch(final String query, @Nullable final ConfluenceUser confluenceUser, final int offset,
+                               final int limit) {
+        Set<SearchQuery> searchQueries = new HashSet<>();
+
+        searchQueries.add(new TextQuery(query));
+        searchQueries.add(new ArchivedSpacesQuery(false, getSpaceManager()));
+
+        if (confluenceUser != null) {
+            searchQueries.add(ContentPermissionsQuery.builder()
+                    .user(confluenceUser)
+                    .build());
+            searchQueries.add(getSpacePermissionQueryFactory().create(confluenceUser));
+        }
+        SearchQuery searchQuery = BooleanQuery.composeAndQuery(searchQueries);
+        RelevanceSort sort = new RelevanceSort();
+        ISearch searchConfig = new ContentSearch(searchQuery, sort, offset, limit);
+
+        return searchConfig;
+    }
+
+    private SpacePermissionQueryFactory getSpacePermissionQueryFactory() {
```

Comment:
Will there will be any issue if we move getSpacePermissionQueryFactory() to  BaseConfluenceCompatibilityHandler class just like we did for getSpaceManager()

Reply:
`BaseConfluenceCompatibilityHandler` is located in a different `confluence-compat-common` module, that I wanted to keep free from product-specific libraries. The reason for it - to make it compatible with all Confluence versions - 7.10 - 8.0. `SpacePermissionQueryFactory` is available only [since Confluence 7.17](https://docs.atlassian.com/ConfluenceServer/javadoc/7.17.0/com/atlassian/confluence/search/v2/SpacePermissionQueryFactory.html) while we need to support older Confluence 7.10 as well. So, if this method is moved to `BaseConfluenceCompatibilityHandler` the class would become compatible only with Confluence 7.17+. That's why this method present in compatibility handler for Confluence 8. We may be sure that `SpacePermissionQueryFactory` will be present in all 8.x versions. It isn't true for 7.x versions unfortunately.